### PR TITLE
implement client credentials oauth flow

### LINKED ISSUE #344: implement client credentials oauth flow
for rstudio/connect/issues/28386, add a function to exchange a content-session-token for a client credential access token

### PR DESCRIPTION
**Pull request by christierney**
## Intent

Add oauth client credential token function to the package.

Fixes rstudio/connect#28386
Fixed #344 

## Approach

There was some discussion at https://github.com/rstudio/connectapi/pull/292#discussion_r1738978267 about whether it would be better to use a different function for client credentials vs viewer auth, or one function with different optional params. I opted for a separate function and followed the naming pattern from https://github.com/posit-dev/posit-sdk-py/pull/348/files

## Manual testing

From RStudio, installed development version of the library:

```r
remotes::install_github("rstudio/connectapi@chris-28386-cred-exchange")
```

and deployed this shiny app to local Connect:

```r
library(connectapi)
library(shiny)

ui <- fluidPage(
  textOutput("content_session_token"),
  textOutput("content_token"),
  )

server <- function(input, output, session) {


  content_session_token <- Sys.getenv("CONNECT_CONTENT_SESSION_TOKEN")
  output$content_session_token <- renderText(content_session_token)

  client <- connect(
    server = "http://localhost:3939",
    api_key = Sys.getenv("CONNECT_API_KEY"),
  )

  output$content_token <- renderText({
    resp <- get_oauth_content_credentials(client, content_session_token)
    resp$access_token
  })
}

# start the Shiny app
shinyApp(ui = ui, server = server)
```

Configured the content to use a valid service account and verified I can get a token:

Screenshot 2024-12-04 at 4 49 12 PM

## Checklist

- [x] Does this change update `NEWS.md` (referencing the connected issue if necessary)?
- [x] Does this change need documentation? Have you run `devtools::document()`?

### CODE REVIEW

File: R/get.R
Comment (christierney) [inline R/get.R:663]: note: this doesn't exist yet but feels like a section we will add? Corresponds to the existing `#obtaining-a-viewer-oauth-access-token` section.

**Review by zackverham — APPROVED** [review]
LGTM but obviously want Toph to take a look still!

File: R/get.R
Comment (toph-allen) [inline R/get.R:636]:
```suggestion
#' @param content_session_token Optional. The content session token. This token can only
```

To go along with suggestion below. And I'd also adjust the prose a little too maybe :)

**Review by toph-allen — CHANGES_REQUESTED** [review]
Looks good, but I think having it automatically retrieve the token from the environment in the default case is nicer — could you make that change?

File: R/get.R
Comment (christierney) [reply R/get.R:669]: thank you! this is super helpful.

File: R/get.R
Comment (zackverham) [reply R/get.R:661]: here's the in-flight draft of the user guide changes: 
https://github.com/rstudio/connect/pull/29116

The parallel actually ended up being `#obtaining-a-service-account-oauth-access-token`

**Review by toph-allen — APPROVED** [review]
Looks good! Just one small non-functional change, so approving. :) If you know you're gonna get that docs link in for this Connect release, I'm happy to merge now, or we can hold off on merging till that happens.

Thinking out loud here: Do you think it's worth renaming `get_oauth_credentials()` to `get_oauth_user_credentials()`? Could do that with a deprecation — keep the existing function around but do a deprecation warning and just have it call the renamed function. That way we have `get_oauth_content_credentials()` and `get_oauth_user_credentials()`. Happy to make that change after this merges in a separate PR.

**Comment by toph-allen** [comment]
It looks like CI is failing because of something flaky, not for an actual reason, but if it continues to fail I'm happy to pull locally and poke around at it.

**Comment by christierney** [comment]
re. naming: we'd also want to change the name in the python SDK which currently follows the same convention as this one. This also follows the convention in the Go oauth package. I do think it's a question worth raising, but probably separate from this PR.

File: R/utils.R
Comment (toph-allen) [inline R/utils.R:159]: I changed the error messages.

- The phrase "this API requires…" doesn't really make sense. Users of the package will be thinking in terms of features or functions.
- The guidance in the error case was confusing. Using a previous version of `connectapi` would not help! If the feature doesn't exist on an old version of Connect, using an older `connectapi` will… just remove the `connectapi` references to that feature. Reviewing the API docs won't help either, because users won't see this error if they are directly calling the API, only if they're using existing functions.

**Comment by christierney** [comment]
@toph-allen the tests don't need to change to accommodate the version check? What version of Connect do tests see?

**Comment by christierney** [comment]
oh wait, toph already explained this to me in slack:

> Ah, the tests don’t fail because the mock server version is NA (i.e. not exposed), which just results in a warning, not an outright error. I thought that might be the case, probably I adjusted it for this very reason at some point

**Review by toph-allen — APPROVED** [review]

**Pull request merged**
